inc/azure_macro_utils/macro_utils.h

```diff
@@ -78,73 +78,101 @@ MU_IF(X, "true", "false") => "true"
-                *destination = (enumName)i;                                     \
-                return 0;                                                       \
-            }                                                                   \
-        }                                                                       \
-        return MU_FAILURE;                                                      \
-    }                                                                           \
-}                                                                               \
+    MU_DEFINE_ENUM_WITHOUT_INVALID(enumName, MU_C2(enumName, _INVALID), __VA_ARGS__) \
+
+#define MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID(enumName, ...) \
+    const char* MU_C3(MU_, enumName, _ToString)(enumName value) \
+    { \
+        MU_FOR_EACH_1(MU_INTERNAL_DECLARE_ENUM_VAR, __VA_ARGS__) \
+        int enum_last_value = -1; \
+        MU_FOR_EACH_1(MU_INTERNAL_ASSIGN_ENUM_VALUE, __VA_ARGS__) \
+        return "NULL"; \
+    }
 
 // this macro is a wrapper on top of MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID, adding an _INVALID value as the first enum value in the enum
 #define MU_DEFINE_ENUM_STRINGS(enumName, ...) \
     MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID(enumName, MU_C2(enumName, _INVALID), __VA_ARGS__)
 
-#define MU_DEFINE_LOCAL_ENUM_WITHOUT_INVALID(enumName, ...) typedef enum MU_C2(enumName, _TAG) { MU_FOR_EACH_1(MU_DEFINE_ENUMERATION_CONSTANT, __VA_ARGS__)} enumName; \
-static const char* MU_C2(enumName, StringStorage)[MU_COUNT_ARG(__VA_ARGS__)] = {MU_FOR_EACH_1(MU_DEFINE_ENUMERATION_CONSTANT_AS_STRING, __VA_ARGS__)}; \
-static const char* MU_C2(enumName,Strings)(enumName value)         \
-{                                                                  \
-    if((int)value<0 || (int)value>=MU_COUNT_ARG(__VA_ARGS__))      \
-    {                                                              \
-        /*this is an error case*/                                  \
-        return "NULL";                                             \
-    }                                                              \
-    else                                                           \
-    {                                                              \
-        return MU_C2(enumName, StringStorage)[value];              \
-    }                                                              \
-}
+#define MU_ENUM_TO_STRING(enumName, value) \
+    MU_C3(MU_, enumName, _ToString)(value)
+
+//
+//
+// MU_DEFINE_LOCAL_ENUM
+//
+//
+
+#define MU_DEFINE_ENUMERATION_CONSTANT_AS_STRING(enumValue) \
+    MU_TOSTRING(enumValue),
+
+#define MU_DEFINE_LOCAL_ENUM_WITHOUT_INVALID(enumName, ...) \
+    typedef enum MU_C2(enumName, _TAG) { MU_FOR_EACH_1(MU_DEFINE_ENUMERATION_CONSTANT, __VA_ARGS__)} enumName; \
+    static const char* MU_C3(MU_, enumName, _ToString)(enumName value) \
+    { \
+        MU_FOR_EACH_1(MU_INTERNAL_DECLARE_ENUM_VAR, __VA_ARGS__) \
+        int enum_last_value = -1; \
```

Comment:
>-1 [](start = 30, length = 2)

what happens when one of the enum values is initialized to the constant expression "-1"?

Reply:
That is all fine, we have a test for that now.

---
In reply to: [328356458](https://github.com/Azure/azure-macro-utils-c/pull/30#discussion_r328356458) [](ancestors = 328356458)